README.md
# Incident Management System

Fullstack Serverless Application on AWS

## Architecture

![](Image.png)

## AWS IAM

```
aws configure --profile mycli
```

## Amazon DynamoDB

```
aws dynamodb create-table \
    --table-name Incidents \
    --attribute-definitions AttributeName=IncidentTitle, AttributeType=S AttributeName=IncidentDate,AttributeType=S,
    --key-schema AttributeName=IncidentTitle,KeyType=HASH AttributeName=IncidentName,KeyType=Range
    --tags Key="Project" Value="Incident Management System"
```

## AWS Lambda

### Create AWS Lambda Functions

```
aws lambda create-function --function-name createIncident \
    --runtime nodejs12.x \
    --role arn:aws:iam:0123456789012:role/DynamoDBFullLambdaAccess \
    --handler createIncident.create \
    --zip-file fileb://createIncident.zip
```

```
aws lambda create-function --function-name getIncident \
    --runtime nodejs12.x \
    --role arn:aws:iam:0123456789012:role/DynamoDBFullLambdaAccess \
    --handler getIncident.get \
    --zip-file fileb://getIncident.zip
```

```
aws lambda create-function --function-name updateIncident \
    --runtime nodejs12.x \
    --role arn:aws:iam:0123456789012:role/DynamoDBFullLambdaAccess \
    --handler updateIncident.update \
    --zip-file fileb://updateIncident.zip
```

```
aws lambda create-function --function-name listIncidents \
    --runtime nodejs12.x \
    --role arn:aws:iam:0123456789012:role/DynamoDBFullLambdaAccess \
    --handler listIncidents.list \
    --zip-file fileb://listIncidents.zip
```

```
aws lambda create-function --function-name deleteIncident \
    --runtime nodejs12.x \
    --role arn:aws:iam:0123456789012:role/DynamoDBFullLambdaAccess \
    --handler deleteIncident.delete \
    --zip-file fileb://deleteIncident.zip
```

### Test Lambda Functions

```
aws dynamodb get-item --table-name Incidents \
    --key '{"IncidentTitle": {"S":"Tornado in Tulsa","IncidentDate":{"S": "2020-07-23"}}'
```

```
aws lambada invoke --function-name updateIncident \
    --payload file://updateTest.json \
    updateOutfile.txt
```

```
aws lambda invoke --function-name listIncidents \
    --payload '{}' \
    --listOutfile.txt
```

```
aws lambada invoke --function-name deleteIncident \
    --payload file://deleteTest.json \
    deleteOutfile.txt
```

## Amazon API Gateway

## Amazon S3
[Development Website](http://djs-incident-management-system.s3-website-us-east-1.amazonaws.com)

## Built With

* [Visual Studio Code](https://code.visualstudio.com) - Source Code Editor
* [Amazon Web Services](https://aws.amazon.com) - Cloud Computing Services
* [Amazon DynamoDB Developer Guide](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GettingStarted.NodeJs.html) - Getting Started with Node.js and DynamoDB

## Authors

* **Daniel J. Sablosky** - *Initial work* - [Portfolio](https://portfolio.djsablosky.info)

## License

This project is licensed under the GNU General Public License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgements

* [Linux Academy](https://linuxacademy.com) - Fullstack Serverless Applications on AWS

CHANGELOG.md
# Incident Management System CHANGELOG

This file is used to list changes made in each version of the fullstack serverless application.

# 0.1.0

Initial  release.

- AWS Lambda functions
- Static site 

formlogic.js
// Replace the YOUR_API_ENDPOINT_URL with yours
// It should look something like this:
// https://example1a2s3d.execute-api.us-east-1.amazonaws.com/dev/incidents

// Do NOT include the /id but make sure you DO include the /incidents

var API_ENDPOINT = 'https://d98w8xrrc1.execute-api.us-east-1.amazonaws.com/dev/incidents';

// Setup divs that will be used to display interactive messages
var errorDiv = document.getElementById('error-message')
var successDiv = document.getElementById('success-message')
var resultsDiv = document.getElementById('results-message')

// Setup easy way to reference values of the input boxes
function TitleValue() { return document.getElementById('Title').value }
function DateValue() { return document.getElementById('Date').value }
function DescriptionValue() { return document.getElementById('Description').value }
function StatusValue() { return document.getElementById('Status').value }
function TypeValue() { return document.getElementById('Type').value }

function clearNotifications() {
    // Clear any exisiting notifications in the browser notifications divs
    errorDiv.textContent = '';
    resultsDiv.textContent = '';
    successDiv.textContent = '';
}

// Add listeners for each button that make the API request

document.getElementById('listButton').addEventListener('click', function (event) {
    // Prevent the page reloading and clear exisiting notifications
    event.preventDefault()
    clearNotifications()
    // Prepare the appropriate HTTP request to the API with fetch
    // list uses the base /prometheon resource path and doesn't require
    // a payload or query string paramaters
    fetch(API_ENDPOINT, {
        method: 'GET',
        mode: 'cors'
    })
    .then((resp) => resp.json()) 
    .then(function(data) {
        console.log(data)
        successDiv.textContent = 'Success! Check the fetched items below';
        resultsDiv.textContent = JSON.stringify(data);
    })
    .catch(function(err) {
        errorDiv.textContent = 'Yikes! There was an error:\n' + err.toString();
        console.log(err)
    });
});


document.getElementById('getButton').addEventListener('click', function (event) {
    // Prevent the page reloading and clear exisiting notifications
    event.preventDefault()
    clearNotifications()
    // Prepare the appropriate HTTP request to the API with fetch
    // get uses the /prometheon/id resource path and query string paramaters
    // The final result is a GET request to something like:
    // https://example1a2s3d.execute-api.us-east-1.amazonaws.com/dev/prometheon/id?Artist=Rihanna&SongTitle=Work
    fetch(API_ENDPOINT+'/id?Title='+TitleValue()+'&Date='+DateValue(), {
        headers:{
            "Content-type": "application/json"
        },
        method: 'GET',
        mode: 'cors'
    })
    .then((resp) => resp.json()) 
    .then(function(data) {
        console.log(data)
        successDiv.textContent = 'Success! Check the fetched item below';
        resultsDiv.textContent = JSON.stringify(data);
    })
    .catch(function(err) {
        errorDiv.textContent = 'Yikes! There was an error:\n' + err.toString();
        console.log(err)
    });
});


document.getElementById('deleteButton').addEventListener('click', function (event) {
    // Prevent the page reloading and clear exisiting notifications
    event.preventDefault()
    clearNotifications()
    // Prepare the appropriate HTTP request to the API with fetch
    // Delete used the /prometheon/id resource path 
    // It also requires a JSON body payload
    fetch(API_ENDPOINT+'/id', {
        headers:{
            "Content-type": "application/json"
        },
        method: 'DELETE',
        body: JSON.stringify({
            'Title': TitleValue(),
            'Date': DateValue()
        }),
        mode: 'cors'
    })
    .then((resp) => resp.json()) 
    .then(function(data) {
        console.log(data)
        successDiv.textContent = 'Success! Check the status of the item below';
        resultsDiv.textContent = JSON.stringify(data);
    })
    .catch(function(err) {
        errorDiv.textContent = 'Yikes! There was an error:\n' + err.toString();
        console.log(err)
    });
});


document.getElementById('updateButton').addEventListener('click', function (event) {
    // Prevent the page reloading and clear exisiting notifications
    event.preventDefault()
    clearNotifications()    
    // Prepare the appropriate HTTP request to the API with fetch
    // update uses the /prometheon/id endpoint and requires a JSON payload
    fetch(API_ENDPOINT+'/id', {
        headers:{
            "Content-type": "application/json"
        },
        method: 'PUT',
        body: JSON.stringify({
            'Title': TitleValue(),
            'Date': DateValue(),
            'Description': DescriptionValue(),
            'Status': StatusValue(),
            'Type': TypeValue()
        }),
        mode: 'cors'
    })
    .then((resp) => resp.json()) 
    .then(function(data) {
        console.log(data)
        successDiv.textContent = 'Success! Check the updated item below';
        resultsDiv.textContent = JSON.stringify(data);

    })
    .catch(function(err) {
        errorDiv.textContent = 'Yikes! There was an error:\n' + err.toString();
        console.log(err)
    });
});


document.getElementById('createButton').addEventListener('click', function (event) {
    // Prevent the page reloading and clear exisiting notifications
    event.preventDefault()
    clearNotifications()
    // Prepare the appropriate HTTP request to the API with fetch
    // create uses the root /prometheon endpoint and requires a JSON payload
    fetch(API_ENDPOINT, {
        headers:{
            "Content-type": "application/json"
        },
        method: 'POST',
        body: JSON.stringify({
            'Title': TitleValue(),
            'Date': DateValue(),
            'Description': DescriptionValue(),
            'Status': StatusValue(),
            'Type': TypeValue()
        }),
        mode: 'cors'
    })
    .then((resp) => resp.json()) 
    .then(function(data) {
        console.log(data)
        successDiv.textContent = 'Success! Check the created item below';
        resultsDiv.textContent = JSON.stringify(data);
    })
    .catch(function(err) {
        errorDiv.textContent = 'Yikes! There was an error:\n' + err.toString();
        console.log(err)
    });
});

Lambda/getIncident.js
'use strict';

const AWS = require('aws-sdk');
AWS.config.update({region:'us-east-1'});

const dynamoDb = new AWS.DynamoDB.DocumentClient();

module.exports.get = (event, context, callback) => {
  const data = event.queryStringParameters;

  const params = {
    TableName: 'Incidents',
    Key: {
      "IncidentTitle": data.IncidentTitle,
      "IncidentDate": data.IncidentDate
    },
  }
  
  // fetch item from DynamoDB
  dynamoDb.get(params, (error, result) => {
    // handle potential errors
    if (error) {
      console.error(error);
      callback(null, {
        statusCode: error.statusCode || 501,
        headers: { 'Content-Type': 'text/plain' },
        body: 'Couldn\'t fetch the todo item.',
      });
      return;
    }

    // create a response
    const response = {
      statusCode: 200,
      headers: {'Access-Control-Allow-Origin': '*'},
      body: JSON.stringify(result),
    };
    callback(null, response);
  });
}

Lambda/listIncidents.js
'use strict';

const AWS = require('aws-sdk');
AWS.config.update({region:'us-east-1'});

const dynamoDb = new AWS.DynamoDB.DocumentClient();

module.exports.list = (event, context, callback) => {

    const params = {
        TableName: 'Incidents',
        Limit: 10
    }

    // fetch items from DynamoDB
    dynamoDb.scan(params, (error, result) => {
        // handle potential errors
        if (error) {
            console.error(error);
            callback(null, {
                statusCode: error.statusCode || 501,
                headers: {'Content-Type': 'text/plain' },
                body: 'Couldn\'t fetch the items.',
            });
            return;
        }

        // create a response
        const response = {
            statusCode: 200,
            headers: {'Access-Control-Allow-Origin': '*'},
            body: JSON.stringify(result),
        };
        callback(null, response);
    });
};

Lambda/createIncident.js
'use strict';

const AWS = require('aws-sdk');
AWS.config.update({region:'us-east-1'});

const dynamoDb = new AWS.DynamoDB.DocumentClient();

module.exports.create = (event, context, callback) => {
  const data = JSON.parse(event.body);

  const params = {
    TableName: 'Incidents',
    Item: {
      IncidentTitle: data.IncidentTitle,
      IncidentDate: data.IncidentDate,
      IncidentDescription: data.IncidentDescription,
      IncidentStatus: data.IncidentStatus,
      IncidentType: data.IncidentType
    },
  };

  // write the item to the database
  dynamoDb.put(params, (error) => {
    // handle potential errors
    if (error) {
      console.error(error);
      callback(null, {
        statusCode: error.statusCode || 501,
        headers: { 'Content-Type': 'text/plain' },
        body: 'Couldn\'t create the item.',
      });
      return;
    }

    // create a response
    const response = {
      statusCode: 200,
      headers: {'Access-Control-Allow-Origin': '*'},
      body: JSON.stringify(params.Item),
    };
    callback(null, response);
  });
}

Lambda/deleteIncident.js
'use strict';

const AWS = require('aws-sdk');
AWS.config.update({region:'us-east-1'});

const dynamoDb = new AWS.DynamoDB.DocumentClient();

module.exports.delete = (event, context, callback) => {
  const data = JSON.parse(event.body);

  const params = {
    TableName: 'Incidents',
    Key: {
      "IncidentTitle": data.IncidentTitle,
      "IncidentDate": data.IncidentDate
    },
  }

  // delete the item from DynamoDB
  dynamoDb.delete(params, (error, result) => {
    // handle potential errors
    if (error) {
      console.error(error);
      callback(null, {
        statusCode: error.statusCode || 501,
        headers: { 'Content-Type': 'text/plain' },
        body: 'Couldn\'t delete the item.',
      });
      return;
    }

    // create a response
    const response = {
      statusCode: 200,
      headers: {'Access-Control-Allow-Origin': '*'},
      body: JSON.stringify('Item Deleted'),
    };
    callback(null, response);
  });
}

Lambda/updateIncident.js
'use strict';

const AWS = require('aws-sdk');
AWS.config.update({region:'us-east-1'});

const dynamoDb = new AWS.DynamoDB.DocumentClient();

module.exports.update = (event, context, callback) => {
  const data = JSON.parse(event.body);

  const params = {
    TableName: 'Incidents',
    Key: {
      "IncidentTitle": data.IncidentTitle,
      "IncidentDate": data.IncidentDate
    },
    UpdateExpression: "set IncidentDescription=:IncidentDescription, IncidentStatus=:IncidentStatus, IncidentType=:IncidentType",
    ExpressionAttributeValues:{
        ":IncidentDescription": data.IncidentDescription,
        ":IncidentStatus": data.IncidentStatus,
        ":IncidentType": data.IncidentType
    },
    ReturnValues:"ALL_NEW"
  };
  
  // write item to DynamoDB
  dynamoDb.update(params, (error) => {
    // handle potential errors
    if (error) {
      console.error(error);
      callback(null, {
        statusCode: error.statusCode || 501,
        headers: { 'Content-Type': 'text/plain' },
        body: 'Couldn\'t create the item.',
      });
      return;
    }

    // create a response
    const response = {
      statusCode: 200,
      headers: {'Access-Control-Allow-Origin': '*'},
      body: JSON.stringify("Item Updated"),
    };
    callback(null, response);
  });
};